src/main/kotlin/com/openlattice/data/controllers/DataController.kt

```diff
@@ -0,0 +1,922 @@
+            @RequestParam(ENTITY_SET_ID) entitySetId: UUID,
+            @RequestBody entities: List<Map<UUID, Set<Any>>>
+    ): List<UUID> {
+        //Ensure that we have read access to entity set metadata.
+        ensureReadAccess(AclKey(entitySetId))
+        ensureEntitySetCanBeWritten(entitySetId)
+        val requiredPropertyTypes = entities.flatMap { entity -> entity.keys }.toSet()
+
+        //Load authorized property types
+        val authorizedPropertyTypes = authzHelper
+                .getAuthorizedPropertyTypes(entitySetId, EdmAuthorizationHelper.WRITE_PERMISSION)
+        accessCheck(authorizedPropertyTypes, requiredPropertyTypes)
+        val entityKeyIdsToWriteEvent = dgm.createEntities(entitySetId, entities, authorizedPropertyTypes)
+        val entityKeyIds = entityKeyIdsToWriteEvent.key
+
+        recordEvent(AuditableEvent(
+                spm.getCurrentUserId(),
+                AclKey(entitySetId),
+                AuditEventType.CREATE_ENTITIES,
+                "Entities created through DataApi.createEntities",
+                Optional.of<Set<UUID>>(Sets.newHashSet(entityKeyIds)),
+                mapOf(),
+                entityKeyIdsToWriteEvent.value.version
+        ))
+
+        return entityKeyIds
+    }
+
+    @Timed
+    @RequestMapping(value = [ENTITY], method = [RequestMethod.POST], consumes = [MediaType.APPLICATION_JSON_VALUE])
+    override fun createEntities(@RequestBody entities: Map<UUID, List<Map<UUID, Set<Any>>>>): Map<UUID, List<UUID>> {
+        return entities.mapValues { (entitySetId, entityData) ->
+            createEntities(entitySetId, entityData)
+        }
+    }
+
+    @Timed
+    @RequestMapping(path = [ASSOCIATION], method = [RequestMethod.POST], consumes = [MediaType.APPLICATION_JSON_VALUE])
+    override fun createAssociations(@RequestBody associations: Map<UUID, List<DataEdge>>): Map<UUID, List<UUID>> {
+        //Ensure that we have read access to entity set metadata.
+        val entitySetIds = getEntitySetIdsFromCollection(associations.values.flatten()) { dataEdge ->
+            listOf(dataEdge.src.entitySetId, dataEdge.dst.entitySetId)
+        }
+        checkPermissionsOnEntitySetIds(entitySetIds, EdmAuthorizationHelper.READ_PERMISSION)
+
+        //Ensure that we can write properties.
+        val requiredPropertyTypes = requiredAssociationPropertyTypes(associations) { it }
+        accessCheck(EdmAuthorizationHelper.aclKeysForAccessCheck(
+                requiredPropertyTypes, EdmAuthorizationHelper.WRITE_PERMISSION
+        ))
+
+        val authorizedPropertyTypesByEntitySet = authzHelper
+                .getAuthorizedPropertiesOnEntitySets(associations.keys, EdmAuthorizationHelper.WRITE_PERMISSION)
+
+        dataGraphServiceHelper.checkAssociationEntityTypes(associations)
+        val associationsCreated = dgm.createAssociations(associations, authorizedPropertyTypesByEntitySet)
+
+        val associationIds = mutableMapOf<UUID, List<UUID>>()
```

Comment:
Can size this with `associationsCreated.size()`

Reply:
👍